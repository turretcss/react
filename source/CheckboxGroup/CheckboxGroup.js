import React from 'react'

import { Checkbox } from '../Control'

const CheckboxGroup = ({ value = [], onChange, options = [], ...props }) =>
  options.map(option => (
    <Checkbox
      {...props}
      key={option.value}
      value={contains(value, option.value)}
      onChange={checked => onChange(toggleOption(value, options, option.value, checked))}
      label={option.label}
    />
  ))

function toggleOption(values, options, optionValue, checked) {
  return options
    .filter(({ value }) => {
      if (optionValue !== value) {
        return contains(values, value)
      } else {
        return checked
      }
    })
    .map(({ value }) => value)
}

function contains(values, value) {
  return Array.isArray(values) && values.indexOf(value) !== -1
}

export default CheckboxGroup
