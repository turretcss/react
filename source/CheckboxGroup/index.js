export { default } from './CheckboxGroup'
export { default as CheckboxGroup } from './CheckboxGroup'
