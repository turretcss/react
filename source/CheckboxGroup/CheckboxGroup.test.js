import React from 'react'
import renderer from 'react-test-renderer'
import { renderToStaticMarkup } from 'react-dom/server'

import CheckboxGroup from './CheckboxGroup'

const options = [
  {
    value: 's',
    label: 'Small',
  },
  {
    value: 'm',
    label: 'Medium',
  },
  {
    value: 'l',
    label: 'Large',
  },
]

describe('CheckboxGroup', () => {
  it('renders nothing if no options supplied', () => {
    const result = renderToStaticMarkup(<CheckboxGroup />)
    const expected = ''
    expect(result).toEqual(expected)
  })

  it('renders array of checkboxes', () => {
    const tree = renderer.create(<CheckboxGroup options={options} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders array of checkboxes with "Small" option checked', () => {
    const tree = renderer.create(<CheckboxGroup value={['s']} options={options} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('renders array of checkboxes with "Small" and "Large" option checked', () => {
    const tree = renderer.create(<CheckboxGroup value={['s', 'l']} options={options} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('passes unknown props down to checkboxes', () => {
    const tree = renderer
      .create(<CheckboxGroup options={options} name="custom-name" size="xs" disabled color="error" />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('spreads option attributes to checkboxes', () => {
    const options = [
      {
        value: 'one',
        label: 'One',
        color: 'success',
      },
    ]
    const tree = renderer.create(<CheckboxGroup options={options} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('calls onChange when checkbox clicked', () => {
    let newValue = null
    const testInstance = renderer.create(<CheckboxGroup onChange={value => (newValue = value)} options={options} />)
      .root
    // click checkbox with value "s"
    testInstance.findAllByType('input')[0].props.onClick({ target: { checked: true } })
    expect(newValue).toEqual(['s'])
  })

  it('toggles value in array when checkbox clicked', () => {
    let newValue = null
    const testInstance = renderer.create(
      <CheckboxGroup value={['s', 'm']} onChange={value => (newValue = value)} options={options} />
    ).root
    // click checkbox with value "s"
    testInstance.findAllByType('input')[0].props.onClick({ target: { checked: false } })
    expect(newValue).toEqual(['m'])
  })

  it('ignores unknown values (ie. not in options list) when clicking checkbox', () => {
    let newValue = null
    const testInstance = renderer.create(
      <CheckboxGroup value={['unknown']} onChange={value => (newValue = value)} options={options} />
    ).root
    // click checkbox with value "s"
    testInstance.findAllByType('input')[0].props.onClick({ target: { checked: true } })
    expect(newValue).toEqual(['s'])
  })
})
