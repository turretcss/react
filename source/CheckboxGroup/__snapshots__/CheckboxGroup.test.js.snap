// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckboxGroup passes unknown props down to checkboxes 1`] = `
Array [
  <label
    className="control checkbox control-xs error"
  >
    <input
      checked={false}
      disabled={true}
      name="custom-name"
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Small
    </span>
  </label>,
  <label
    className="control checkbox control-xs error"
  >
    <input
      checked={false}
      disabled={true}
      name="custom-name"
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Medium
    </span>
  </label>,
  <label
    className="control checkbox control-xs error"
  >
    <input
      checked={false}
      disabled={true}
      name="custom-name"
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Large
    </span>
  </label>,
]
`;

exports[`CheckboxGroup renders array of checkboxes 1`] = `
Array [
  <label
    className="control checkbox"
  >
    <input
      checked={false}
      disabled={undefined}
      name={undefined}
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Small
    </span>
  </label>,
  <label
    className="control checkbox"
  >
    <input
      checked={false}
      disabled={undefined}
      name={undefined}
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Medium
    </span>
  </label>,
  <label
    className="control checkbox"
  >
    <input
      checked={false}
      disabled={undefined}
      name={undefined}
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Large
    </span>
  </label>,
]
`;

exports[`CheckboxGroup renders array of checkboxes with "Small" and "Large" option checked 1`] = `
Array [
  <label
    className="control checkbox"
  >
    <input
      checked={true}
      disabled={undefined}
      name={undefined}
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Small
    </span>
  </label>,
  <label
    className="control checkbox"
  >
    <input
      checked={false}
      disabled={undefined}
      name={undefined}
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Medium
    </span>
  </label>,
  <label
    className="control checkbox"
  >
    <input
      checked={true}
      disabled={undefined}
      name={undefined}
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Large
    </span>
  </label>,
]
`;

exports[`CheckboxGroup renders array of checkboxes with "Small" option checked 1`] = `
Array [
  <label
    className="control checkbox"
  >
    <input
      checked={true}
      disabled={undefined}
      name={undefined}
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Small
    </span>
  </label>,
  <label
    className="control checkbox"
  >
    <input
      checked={false}
      disabled={undefined}
      name={undefined}
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Medium
    </span>
  </label>,
  <label
    className="control checkbox"
  >
    <input
      checked={false}
      disabled={undefined}
      name={undefined}
      onChange={[Function]}
      onClick={[Function]}
      type="checkbox"
    />
    <span
      className="control-indicator"
    />
    <span
      className="control-label"
    >
      Large
    </span>
  </label>,
]
`;
